ivy/func_wrapper.py

```diff
@@ -1181,11 +1185,23 @@ def method(*args, **kwargs):
                 # no unsupported dtype specified
                 return fn(*args, **kwargs)
 
+        # specifies which dtype to cast the output to
+        to_cast = None
         if "dtype" in kwargs and kwargs["dtype"] is not None:
+            to_cast = kwargs["dtype"]
             dtype = caster(kwargs["dtype"], intersect)
             if dtype:
                 kwargs["dtype"] = ivy.as_native_dtype(dtype)
 
+        if not to_cast and ret_dtype_target:
+            for arg in ret_dtype_target:
+                to_cast = ivy.promote_types(
+                    to_cast,
+                    ivy.dtype(
+                        args[arg_names.index(arg)] if arg not in kwargs else kwargs[arg]
+                    ),
+                )
```

Comment:
this does not take into consideration when the inputs are python scalars or other objects that do not have a `dtype` attribute, we should stick to using `promote_types_of_inputs` to guarantee consistent results, also to prevent the existence of two identical promotion codes which will double the maintenance efforts

Reply:
another problem is that this doesn't take into consideration default values that are optional but still affect the output dtype

Reply:
Hey @MahmoudAshraf97 I remember having this conversation a couple of days back but don't fully remember the specifics, could you please share an example where this would be applicable? Thanks and apologies 😅

Reply:
We do have promotion rules for scalars included in `ivy.promote_types_of_inputs` but not in the promotion table hence they are promoted to the default dtype in `ivy.promote_types` which will lead to an inconsistent promotion results

`ivy.promote_types(float_scalar, float16) = float32`
`ivy.promote_types_of_inputs(float_scalar, float16) = float16`

Reply:
Actually `ivy.promote_types` should always receive dtypes (`ivy.Dtype` or `ivy.NativeDtype`), so ideally the result from `ivy.promote_types_of_inputs` is the one which would be used anyway. So ideally `float16` would be the promoted dtype as a result 😄